src/azure-cli-core/azure/cli/core/context.py

```diff
@@ -0,0 +1,158 @@
+        super(ContextNotFoundException, self).__init__(context_name)
+        self.context_name = context_name
+    def __str__(self):
+        return "The context '{}' does not exist.".format(self.context_name)
+
+class ContextExistsException(Exception):
+    def __init__(self, context_name):
+        super(ContextExistsException, self).__init__(context_name)
+        self.context_name = context_name
+    def __str__(self):
+        return "The context '{}' already exists.".format(self.context_name)
+
+class CannotDeleteActiveContextException(Exception):
+    def __str__(self):
+        return "You cannot delete the current active context"
+
+class CannotDeleteDefaultContextException(Exception):
+    def __str__(self):
+        return "You cannot delete the default context."
+
+ACTIVE_CONTEXT_FILE = os.path.join(GLOBAL_CONFIG_DIR, 'active_context')
+ACTIVE_CONTEXT_ENV_VAR = os.environ.get('AZURE_CONTEXT', None)
+DEFAULT_CONTEXT_NAME = 'default'
+
+get_active_context = lambda: get_context(get_active_context_name())
+context_exists = lambda n: any([x for x in get_context_names() if x == n])
+get_context_file_path = lambda n: os.path.join(CONTEXT_CONFIG_DIR, n)
+get_contexts = lambda: [get_context(context_name) for context_name in get_context_names()]
+
+def _set_active_subscription(context):
+    ''' Set the active subscription used by Profile '''
+    from azure.cli.core._profile import Profile, _ENVIRONMENT_NAME, _SUBSCRIPTION_ID, _STATE
+    profile = Profile()
+    subscriptions = profile.load_cached_subscriptions()
+    context_file_path = get_context_file_path(context['name'])
+    context_config = configparser.SafeConfigParser()
+    context_config.read(context_file_path)
+    subscription_to_use = None
+    try:
+        subscription_to_use = context_config.get('context', 'default_subscription')
+    except (configparser.NoSectionError, configparser.NoOptionError):
+        # Fallback to finding the first subscription that is for this cloud
+        for sub in subscriptions:
+            if sub[_ENVIRONMENT_NAME] == context['cloud'] and sub[_STATE] == 'Enabled':
+                subscription_to_use = sub[_SUBSCRIPTION_ID]
+                break
+    if subscription_to_use:
+        profile.set_active_subscription(subscription_to_use)
+
+def get_active_context_name():
+    if ACTIVE_CONTEXT_ENV_VAR:
+        return ACTIVE_CONTEXT_ENV_VAR
+    try:
+        with open(ACTIVE_CONTEXT_FILE, "r") as active_context_file:
+            return active_context_file.read()
+    except (OSError, IOError):
+        return DEFAULT_CONTEXT_NAME
+
+def set_active_context(context_name, skip_set_active_subsciption=False):
+    if not context_exists(context_name):
+        raise ContextNotFoundException(context_name)
+    with open(ACTIVE_CONTEXT_FILE, "w") as active_context_file:
+        active_context_file.write(context_name)
+    if not skip_set_active_subsciption:
+        _set_active_subscription(get_active_context())
+
+def get_context_names():
```

Comment:
I like that "context" isn't "environment"... but context seems more vague and at least as confusing. Why not "settings"? 

When I worked with Android, they use context all over the place, and I still to this day cannot really tell you what the hell a context is in the Android world (an instance of a THING is pretty close but it's very unintuitive).